_posts/2019-11-01-deploying-debezium-with-kafkaconnector-resource.md

```diff
@@ -0,0 +1,697 @@
+┌────────────────────────────────┐         ┌────────────────────────────────┐
+│ minikube, namespace: kafka     │         │ docker                         │
+│                                │         │                                │
+│ Kafka                          │         │ MySQL                          │
+│  name: my-cluster              │         │                                │
+│                                │         └────────────────────────────────┘
+│ KafkaConnect                   │
+│  name: my-connect              │
+│                                │
+└────────────────────────────────┘
+</code></pre>
+
+> Don't worry, the ASCII art gets a lot better than this!
+
+What's missing from this picture is Kafka Connect in the minikube box.
+
+# Kafka Connect image
+
+The next step is to [create a Strimzi Kafka Connect image](https://strimzi.io/docs/master/#creating-new-image-from-base-str) with the Debezium connector jar file. 
+
+First download and extract the Debezium MySQL connector archive
+```shell
+curl https://repo1.maven.org/maven2/io/debezium/debezium-connector-mysql/0.10.0.Final/debezium-connector-mysql-0.10.0.Final-plugin.tar.gz \
+| tar xvz
+```
+
+Prepare a `Dockerfile` which adds those connector files to the Strimzi Kafka Connect image
+
+```shell
+cat <<EOF >Dockerfile
+FROM strimzi/kafka:latest-kafka-2.3.0
+USER root:root
+EXEC mkdir -p /opt/kafka/plugins/debezium
+COPY ./debezium-connector-mysql/ /opt/kafka/plugins/debezium/
+USER 1001
+EOF
+```
+
+Then build the image from that `Dockerfile` and push it to dockerhub.
+
+```shell
+# You can use your own dockerhub organization
+export DOCKER_ORG=tjbentley
+docker build . -t ${DOCKER_ORG}/connect-debezium
+docker push ${DOCKER_ORG}/connect-debezium
+```
+
+
+# Create the Connect cluster
+
+Now we can create a `KafkaConnect` cluster in Kubernetes:
+
+```shell
+cat <<EOF | kubectl apply -f -
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+metadata:
+  name: my-connect-cluster
+  annotations:
+  # use-connector-resources configures this KafkaConnect
+  # to use KafkaConnector resources to avoid
+  # needing to call the Connect REST API directly
+    strimzi.io/use-connector-resources: "true"
+spec:
+  image: ${DOCKER_ORG}/connect-debezium
+  replicas: 1
+  bootstrapServers: my-cluster-kafka-bootstrap:9093
+  tls:
+    trustedCertificates:
+      - secretName: my-cluster-cluster-ca-cert
+        certificate: ca.crt
+EOF
+```
+
+It's worth pointing out a couple of things about the above resource:
+
+* The `strimzi.io/use-connector-resources: "true"` annotation tells the cluster operator that this `KafkaConnect` is managed via `KafkaConnector` resources.
+* The `spec.image` is the image we just created with `docker`.
+
+
+# Create the connector
+
+The last piece is to create the `KafkaConnector` resource configured to connect to our "inventory" database in MySQL.
+
+Here's what the `KafkaConnector` resource looks like:
+
+
+```shell
```

Comment:
Nice! Perhaps also show the password secret stuff we were discussing. That'd add some more meat on the actual new CRD feature this post is about and it nicely matches the DBZ use case.

Reply:
I'm so stupid :disappointed:. I guess this is the point where I _finally_ figure out _exactly_ why we shouldn't use the map-of-key-to-value approach which we've been flip-flopping over the whole time: Although we can mount a Secret into the container in the `KafkaConnect`, we still need to actually configure that parameter in the `KafkaConnector`. If it's mounted as a file we need a way to say "take the value from this file" and if it was set as an env var we still need a way to say "take the value from this env var". So I guess I need to change the `KafkaConnectorSpec` to support this before we merge the PR, right @scholzj @ppatierno ? So the only question is whether it should be a list of 

```
key:
value: #or valueFrom
```

Or a map with parameter name as key and value being an object with `value` and `valueFrom` properties. Since the list approach is consistent with Kubernetes, I propose to do it that way.

Reply:
When we had this discussion I was more for having the consistent Kubernetes approach anyway, even if tbh I didn't recall as you the problem about mounting value from :-) Anyway I am for the list approach.